### PR TITLE
Fix slice range does not overflow available data

### PR DESCRIPTION
**Pull request by rbuffat**
This is a fix for https://github.com/Toblerity/Fiona/issues/813

Currently, we do not test if slices overflow the available data range. Depending on the driver, this results in different behavior. 

This PR adds a check if the start of a slices is within the feature count, and adapts it if not.
In these cases, it is important to have the number of features available. Thus, we need to force GetFeatureCount for Layers that do not support OLC_FASTFEATURECOUNT. This can bring a substantial performance penalty (potential full table scan).

### CODE REVIEW

File: tests/test_non_counting_layer.py [hunk 1]
```diff
@@ -33,5 +33,5 @@ def test_slice(self):
             assert len(features) == 3
 
     def test_warn_slice_negative_index(self):
-        with pytest.warns(RuntimeWarning):
+        with pytest.warns((FionaDeprecationWarning, RuntimeWarning)) as record:
```
Comment (rbuffat) [inline tests/test_non_counting_layer.py:36]: Stacking two pytest.warns does not work, but this does.

**Comment by rbuffat** [comment]
I had a look at the coverage report using https://github.com/Toblerity/Fiona/pull/885 and modified the slice test to cover more cases. Thereby I noticed that the GeoJSON driver for gdal 2.1-2.2 returns the last feature if the index is set by OGR_L_SetNextByIndex to an index greater than the feature count.
Debug output can be seen here: https://travis-ci.org/github/rbuffat/Fiona/jobs/681455774

I fixed that by adding a check if next_index is greater ftcount. However, this fix requires that the feature count does not change over the lifetime of an Iterator. This is a new assumption. GDAL documentation says for OGR_L_GetFeatureCount: `Returns the number of features in the layer. For dynamic databases the count may not be exact.`

**Comment by rbuffat** [comment]
"Just one last thing" I noticed that FeatureBuilder in __next__ of Iterator is wrapped in a try/finally while it isn't in __next__ of ItemsIterator. I stumbled on it because originally cython coverage didn't mark the finally block as covered, but it is executed. This is maybe a bug in cython coverage: https://github.com/cython/cython/issues/3567

**Comment by sgillies** [comment]
@rbuffat thank you!

**Pull request merged**